### PR TITLE
Feature/profile

### LINKED ISSUE #47: Implement Profile View

### LINKED ISSUE #73: Profile view local user info
A local user should be able to update it's info like first name last name and email. There already is a component for this but needs the backend connection still. Might be dependent on another fix of the update user backend

### PR DESCRIPTION
**Pull request by JorisParee**
pin, password and api key are working in profile view. local user information lacks but has a issue.

Fixes #47.
Fixes #73

### CODE REVIEW

**Comment by CodeNamedRobin** [comment]
Noted to fix #47.

File: src/components/ChangeApiKeyComponent.vue [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+function updateApiKey() {
+  authStore.updateUserKey(apiService).then((res) => {
+    //Succes
+    if(res) {
+      toast.add({ severity: "success", summary: "Success", detail: `${t('profile.Your new key is')} \n ${res.key} \n ${t('profile.Kay is not saved')}` });
+    }
+  }).catch((err) => {
+    //error
+    console.log(err);
```
Comment (CodeNamedRobin) [inline src/components/ChangeApiKeyComponent.vue]: Please either wait for the merge of #74 and then implement that error handling here or already merge #74 into your own branch to use that error handling.

File: src/components/ChangeApiKeyComponent.vue [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+const apiKeyMessage = ref();
+const toast = useToast();
+const { t } = useI18n();
+
+
+function updateApiKey() {
+  authStore.updateUserKey(apiService).then((res) => {
+    //Succes
+    if(res) {
+      toast.add({ severity: "success", summary: "Success", detail: `${t('profile.Your new key is')} \n ${res.key} \n ${t('profile.Kay is not saved')}` });
```
Comment (CodeNamedRobin) [inline src/components/ChangeApiKeyComponent.vue]: Formatting.

File: src/components/ChangeApiKeyComponent.vue [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    //error
+    console.log(err);
+  });
+}
+
+
+</script>
+
+<template>
+  <card-component :header="t('profile.Get api key')">
```
Comment (CodeNamedRobin) [inline src/components/ChangeApiKeyComponent.vue]: Please use indices that are camelCase and shorter. The content should not be the key.

File: src/components/ChangeApiKeyComponent.vue [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    <Toast/>
+    <div>
+      <p>{{ $t('profile.Note old api key will get removed') }}</p>
```
Comment (CodeNamedRobin) [inline src/components/ChangeApiKeyComponent.vue]: Idem.

File: src/components/ChangeApiKeyComponent.vue [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      <p>{{apiKeyMessage || '&nbsp;'}}</p>
+    </div>
+    <div>
+      <Button severity="danger" :label="t('profile.Get new API key')" @click="updateApiKey" />
```
Comment (CodeNamedRobin) [inline src/components/ChangeApiKeyComponent.vue]: Idem.

File: src/components/ChangeInfoComponent.vue [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  isLocal = (userStore.getCurrentUser.user.type == "LOCAL_USER");
+}
+
+
+const { value: inputFirstName, errorMessage: inputFirstNameError } = useField('inputFirstName', validateFirstName);
+const { value: inputLastName, errorMessage: inputLastNameError } = useField('inputLastName', validateLastName);
+const { value: inputEmail, errorMessage: inputEmailError } = useField('inputEmail', validateEmail);
+
+
+function validateFirstName(checkFirstName: string) {
```
Comment (CodeNamedRobin) [inline src/components/ChangeInfoComponent.vue]: For this we don't need custom validators. See [yup](https://github.com/jquense/yup#schema-defaults).

File: src/components/ChangeInfoComponent.vue [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  if (checkFirstName.length <1) {
+    return "name cannot be empty";
+  }
+  return true;
+}
+
+function validateLastName(checkLastName: string) {
```
Comment (CodeNamedRobin) [inline src/components/ChangeInfoComponent.vue]: Idem.

File: src/components/ChangeInfoComponent.vue [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  if (checkLastName.length <1) {
+    return "name cannot be empty";
+  }
+  return true;
+}
+
+function validateEmail(checkEmail: string) {
```
Comment (CodeNamedRobin) [inline src/components/ChangeInfoComponent.vue]: Idem.

File: src/components/ChangeInfoComponent.vue [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+  if (!checkEmail.match(/^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/)) {
+    return "no valid email";
+  }
+  return true;
+}
+
+function changeUserInfo() {
+  if(validateEmail(inputEmail.value) && validateLastName(inputLastName.value) && validateFirstName(inputFirstName.value)) {
+    //TODO update info
```
Comment (CodeNamedRobin) [inline src/components/ChangeInfoComponent.vue]: Reference the issue or fix.

File: src/components/ChangeInfoComponent.vue [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+  } else {
+    toast.add({ severity: "error", summary: "failed", detail: 'fill in correct info', life: 3000 });
```
Comment (CodeNamedRobin) [inline src/components/ChangeInfoComponent.vue]: Formatting.

File: src/components/ChangeInfoComponent.vue [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+  }
+}
+
+
+</script>
+
+<template>
+  <card-component :header="'Change user info'">
+    <div>
+      <small v-if="!isLocal">!This account is not managed trhoug SudoSOS!</small>
```
Comment (CodeNamedRobin) [inline src/components/ChangeInfoComponent.vue]: Hardcoded text and spelling mistake.

File: src/components/ChangeInfoComponent.vue [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    toast.add({ severity: "error", summary: "failed", detail: 'fill in correct info', life: 3000 });
+
+  }
+}
+
+
+</script>
+
+<template>
+  <card-component :header="'Change user info'">
```
Comment (CodeNamedRobin) [inline src/components/ChangeInfoComponent.vue]: Hardcoded text.

File: src/components/ChangeInfoComponent.vue [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    <div>
+      <small v-if="!isLocal">!This account is not managed trhoug SudoSOS!</small>
+      <div>
+        <p>FirstName</p>
```
Comment (CodeNamedRobin) [inline src/components/ChangeInfoComponent.vue]: Hardcoded text

File: src/components/ChangeInfoComponent.vue [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        <Password :disabled="!isLocal"  v-model="inputFirstName"/>
+        <small class="warning">{{inputFirstNameError || '&nbsp;'}}</small>
+      </div>
+      <div>
+        <p>Last name</p>
```
Comment (CodeNamedRobin) [inline src/components/ChangeInfoComponent.vue]: Hardcoded text.

File: src/components/ChangeInfoComponent.vue [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        <InputText :disabled="!isLocal"  v-model="inputLastName"/>
+        <small class="warning">{{inputLastNameError || '&nbsp;'}}</small>
+      </div>
+      <div>
+        <p>Email adress</p>
```
Comment (CodeNamedRobin) [inline src/components/ChangeInfoComponent.vue]: Hardcoded text.

File: src/components/ChangeInfoComponent.vue [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+        <InputText :disabled="!isLocal" v-modal="inputEmail"/>
+        <small class="warning">{{inputEmailError || '&nbsp;'}}</small>
+      </div>
+      <div>
+        <Button severity="danger" :disabled="true" label="Update user info" @click="changeUserInfo"/>
```
Comment (CodeNamedRobin) [inline src/components/ChangeInfoComponent.vue]: Hardcoded text.

File: src/components/ChangeInfoComponent.vue [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+      </div>
+    </div>
+  </card-component>
+  <Toast />
```
Comment (CodeNamedRobin) [inline src/components/ChangeInfoComponent.vue]: This toast is not needed anymore after mergin #74. Just to keep in mind.

File: src/components/ChangePasswordComponent.vue [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+
+
+//TODO make password greyed when using m account
+const { value: inputPassword, errorMessage: inputPasswordError } = useField('inputPassword', validatePassword);
+const { value: confirmPassword, errorMessage: confirmPasswordError } = useField('confirmPassword', validateConfirmPassword);
+
+
+function validatePassword(checkPassword: string) {
+  if (checkPassword.length <8) {
+    //TODO set password requirements
```
Comment (CodeNamedRobin) [inline src/components/ChangePasswordComponent.vue]: Make issue or fix.

File: src/components/ChangePasswordComponent.vue [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+  isLocal = (userStore.getCurrentUser.user.type == "LOCAL_USER");
+}
+
+
+//TODO make password greyed when using m account
+const { value: inputPassword, errorMessage: inputPasswordError } = useField('inputPassword', validatePassword);
+const { value: confirmPassword, errorMessage: confirmPasswordError } = useField('confirmPassword', validateConfirmPassword);
+
+
+function validatePassword(checkPassword: string) {
```
Comment (CodeNamedRobin) [inline src/components/ChangePasswordComponent.vue]: Custom validators are not needed, as per earlier comment.

File: src/components/ChangePasswordComponent.vue [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+  if (checkPassword.length <8) {
+    //TODO set password requirements
+    return t('profile.Password should');
+  }
+  return true;
+}
+
+function validateConfirmPassword(checkConfirmPassword: string) {
```
Comment (CodeNamedRobin) [inline src/components/ChangePasswordComponent.vue]: Custom validators are not needed.

File: src/components/ChangePasswordComponent.vue [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+function changePassword() {
+  if (validatePassword(inputPassword.value) == true && validateConfirmPassword(confirmPassword.value) == true) {
+    authStore.updateUserLocalPassword(inputPassword.value, apiService).then(() => {
+      inputPassword.value = "";
+      confirmPassword.value = "";
+      toast.add({ severity: "success", summary: "success", detail: "" });
+    }).catch((err) => {
+      console.error(err);
```
Comment (CodeNamedRobin) [inline src/components/ChangePasswordComponent.vue]: Again #74.

File: src/components/ChangePasswordComponent.vue [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+    });
+  } else {
+    toast.add({ severity: "error", summary: "failed", detail: t('profile.Incorrect passwords submitted'), life: 3000 });
```
Comment (CodeNamedRobin) [inline src/components/ChangePasswordComponent.vue]: Hardcoded text in summary.

File: src/components/ChangePasswordComponent.vue [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+  }
+}
+
+</script>
+
+<template>
+  <card-component :header="t('profile.Change password')">
+    <div>
+      <div>
+        <p>Password</p>
```
Comment (CodeNamedRobin) [inline src/components/ChangePasswordComponent.vue]: Hardcoded text.

File: src/components/ChangePasswordComponent.vue [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+        <Password :disabled="!isLocal"  v-model="inputPassword"/>
+        <small class="warning">{{inputPasswordError || '&nbsp;'}}</small>
+      </div>
+      <div>
+        <p>Confirm password</p>
```
Comment (CodeNamedRobin) [inline src/components/ChangePasswordComponent.vue]: Hardcoded text.

File: src/components/ChangeApiKeyComponent.vue (hunk 2 above)
Comment (CodeNamedRobin) [inline src/components/ChangeApiKeyComponent.vue]: Hardcoded text in summary, fixed when #74 is merged.

File: src/components/ChangePinComponent.vue [hunk 25]
```diff
@@ -0,0 +1,82 @@
...
+const { t } = useI18n();
+
+
+const { value: inputPin, errorMessage: inputPinError } = useField('inputPin', validatePin);
+const { value: confirmPin, errorMessage: confirmPinError } = useField('confirmPin', validateConfirmPin);
+
+
+
+//show warning message if pin is filled in and not 4 digits
+function validatePin(checkInputPin: string){
```
Comment (CodeNamedRobin) [inline src/components/ChangePinComponent.vue]: Custom validators.

File: src/components/ChangePinComponent.vue [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+  if (checkInputPin.length != 4 && !RegExp('\\d{4}').test(checkInputPin)) {
+    return t('profile.Pin code must be 4 digits');
+  }
+  return true;
+};
+
+//show warning message if conformation pin does not fit pin
+function validateConfirmPin(checkConfirmPin: string){
```
Comment (CodeNamedRobin) [inline src/components/ChangePinComponent.vue]: Idem.

File: src/components/ChangePinComponent.vue [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+
+function changePinCode() {
+  if (validatePin(inputPin.value) == true && validateConfirmPin(confirmPin.value) == true) {
+    authStore.updateUserPin(inputPin.value, apiService).then(() => {
+      //   Succes!
+      inputPin.value="";
+      confirmPin.value="";
+      toast.add({ severity: "success", summary: "success", detail: t('profile.Pin code updated')  });
+    }).catch((err) => {
+      // Error
```
Comment (CodeNamedRobin) [inline src/components/ChangePinComponent.vue]: Error handling, again #74.

File: src/components/ChangePinComponent.vue [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+
+function changePinCode() {
+  if (validatePin(inputPin.value) == true && validateConfirmPin(confirmPin.value) == true) {
+    authStore.updateUserPin(inputPin.value, apiService).then(() => {
+      //   Succes!
+      inputPin.value="";
+      confirmPin.value="";
+      toast.add({ severity: "success", summary: "success", detail: t('profile.Pin code updated')  });
```
Comment (CodeNamedRobin) [inline src/components/ChangePinComponent.vue]: Hardcoded text in summary.

File: src/components/ChangePinComponent.vue [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+    }).catch((err) => {
+      // Error
+      console.error(err);
+    });
+  } else {
+    toast.add({ severity: "error", summary: "failed", detail: t('profile.Incorrect pin code submitted'), life: 3000 });
```
Comment (CodeNamedRobin) [inline src/components/ChangePinComponent.vue]: Hardcoded text in summary.

File: src/components/ChangePinComponent.vue [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+  }
+}
+
+</script>
+
+<template>
+  <card-component :header="$t('profile.Change pin code')">
+    <Toast />
```
Comment (CodeNamedRobin) [inline src/components/ChangePinComponent.vue]: Again, toast not needed if #74 is merged.

File: src/locales/en.json
Comment (CodeNamedRobin) [inline src/locales/en.json:1]: I already mentioned this, but the keys should be in camelCase and be short and to the point. Actual content should be in the values.

Comment (CodeNamedRobin) [inline src/locales/en.json:1]: Are all these added keys also in nl.json?

File: src/views/ProfileView.vue
Comment (CodeNamedRobin) [inline src/views/ProfileView.vue:1]: Can you format the content in <template> tag? There is a lot of white space for no apparent reason.

File: src/views/ProfileView.vue [hunk 31]
```diff
@@ -0,0 +1,62 @@
...
+
+<script setup lang="ts">
+import ChangePasswordComponent from "@/components/ChangePasswordComponent.vue";
+import ChangePinComponent from "@/components/ChangePinComponent.vue";
+import ChangeApiKeyComponent from "@/components/ChangeApiKeyComponent.vue";
+import CardComponent from "@/components/CardComponent.vue";
+import ChangeInfoComponent from "@/components/ChangeInfoComponent.vue";
+import apiService from "@/services/ApiService";
+
+apiService.user.updateUser(0, {
```
Comment (CodeNamedRobin) [inline src/views/ProfileView.vue]: ??

**Review by CodeNamedRobin — CHANGES_REQUESTED** [review]
There are some things that need to be fixed, both in code style and in aesthetics. I also wonder if this layout of the cards is the best. It right now feels like it is crammed to the left side.

File: src/components/ChangePasswordComponent.vue [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+
+let isLocal = false;
+if(userStore.getCurrentUser.user) {
+  isLocal = (userStore.getCurrentUser.user.type == "LOCAL_USER");
+}
+
+const changeUserPassword = handleSubmit(async (values) => {
+  if (userStore.getCurrentUser.user) {
+    apiService.user.updateUserLocalPassword(
+      userStore.getCurrentUser.user?.id, {password: values.password}).then(() => {
```
Comment (JustSamuel) [inline src/components/ChangePasswordComponent.vue]: missing a .catch

**Review by JustSamuel — CHANGES_REQUESTED** [review]
Only a single mistake

File: src/components/ChangePasswordComponent.vue (hunk 32 above)
Comment (CodeNamedRobin) [reply src/components/ChangePasswordComponent.vue]: Fixed

**Review by CodeNamedRobin — APPROVED** [review]
LGTM!

**Comment by CodeNamedRobin** [comment]
Fixes #73

**Review by JustSamuel — APPROVED** [review]
LGTM

**Pull request merged**